src/components/cuda_pw/linux-cuda.c

```diff
@@ -0,0 +1,377 @@
+        *event_code = evt_rec->evt_code;
+        return PAPI_OK;
+    }
+    else {
+        _papi_hwi_lock(COMPONENT_LOCK);
+        *event_code = global_event_names.count;
+        res = insert_event_record(&global_event_names, name, global_event_names.count, 0);
+        _papi_hwi_unlock(COMPONENT_LOCK);
+    }
+
+    return PAPI_OK;
+}
+
+static int cuda_ntv_code_to_name(unsigned int event_code, char *name, int len)
+{
+    check_n_initialize();
+    if (event_code >= global_event_names.count) {
+        return PAPI_EINVAL;
+    }
+
+    strncpy(name, global_event_names.evts[event_code].name, len);
+    return PAPI_OK;
+}
+
+static int cuda_ntv_code_to_descr(unsigned int event_code, char *descr, int len)
+{
+    check_n_initialize();
+    // strcpy(descr, metric_table->metrics[event_code].description);
+    (void) event_code; (void) descr; (void) len;
+    return PAPI_OK;
+}
+
+static int cuda_init_thread(hwd_context_t *ctx)
+{
+    (void) ctx;
+    return PAPI_OK;
+}
+
+static int cuda_shutdown_thread(hwd_context_t *ctx)
+{
+    (void) ctx;
+    return PAPI_OK;
+}
+
+static int cuda_init_control_state(hwd_control_state_t *ctl)
+{
+    COMPDBG("Entering.\n");
+    (void) ctl;
+    return PAPI_OK;
+}
+
+static int cuda_set_domain(hwd_control_state_t * ctrl, int domain)
+{
+    COMPDBG("Entering\n");
+    (void) ctrl;
+    if((PAPI_DOM_USER & domain) || (PAPI_DOM_KERNEL & domain) || (PAPI_DOM_OTHER & domain) || (PAPI_DOM_ALL & domain))
+        return (PAPI_OK);
+    else
+        return (PAPI_EINVAL);
+    return (PAPI_OK);
+}
+
+static int cuda_update_control_state(hwd_control_state_t *ctl,
+                                     NativeInfo_t *ntv_info,
+                                     int ntv_count, hwd_context_t *ctx
+) {
+    COMPDBG("Entering with events_count %d.\n", ntv_count);
+    check_n_initialize();
+    int i, res;
+    if (ntv_count == 0)
+        goto fn_exit;
+    struct cuda_ctl *control = (struct cuda_ctl *) ctl;
+    struct cuda_ctx *this = (struct cuda_ctx *) ctx;
+    if (this->thread_status == NOT_THR_CURRENT) {
+        this->thread_status = THR_CURRENT;
+        LOCKDBG("Locking.\n");
+        _papi_hwi_lock(_cuda_pw_lock);
+        LOCKDBG("Locked.\n");
+    }
+    if (control->cuctx_arr == NULL) {
+        res = cupti_init_cuctx_arr(&(control->cuctx_arr));
+    }
+    control->events_count = ntv_count;
+
+    if (ntv_count > PAPI_CUDA_MAX_COUNTERS) {
+        ERRDBG("Too many events added.\n");
+        goto fn_fail;
+    }
+    // Add all event names for each gpu in control state
+    for (i=0; i<ntv_count; i++) {
+        control->events_id[i] = ntv_info[i].ni_event;  // store the event code added
+        ntv_info[i].ni_position = i;  // store the mapping of added event index
+    }
+
+    // Validate the added names so far in a temporary context
+    void *tmp_context;
+    res = cupti_control_create(&global_event_names, control->events_count, &(control->events_id[0]), &tmp_context, &(control->cuctx_arr));
```

Comment:
Can you just call `cupti_control_open` and then `cupti_control_close`? I get you have simply changed `open` to `create`. I would keep `open` for consistency with the rocm components.

Reply:
I can do that, but when I was refactoring I thought very hard about this and decided that contexts can be opened and closed, but control cant be opened and closed.

When it comes to the rocm component, are you sure context is the right word? PAPI components already have control and context, they are different things. In the rocm component the control becomes context.

Reply:
Yes I agree control open and close are not ideal. I liked the context term because it is intuitive. We overload both control and context so I don't know why context is suddenly confusing while control is not. I think of context as anything needed by the lower layer to do its job. This contains the events, flags, state, and so on. Context is a fairly used term for this. Control feel more like control state.

Reply:
cuda context is already context, then there is the papi component thread context. there are already too many "context" being used. Need a better way to disambiguate between all three concepts of "context". Since papi component already calls the profiler state control state, we should call it "control" at the low level.